Added configure and make tutorial (#163)
NBISweden/Training-Tech-shorts: posts/2026-04-23-configure-and-make/index.qmd

The good news is that you don't need root access to compile and install software! You just need to tell the installer to put the files somewhere you own, like your home directory or a project directory.

## Background

To follow along in these steps, clone the [Training-Tech-shorts](https://github.com/NBISweden/Training-Tech-shorts) repo to get the example files:

```bash
git clone https://github.com/NBISweden/Training-Tech-shorts
cd Training-Tech-shorts/posts/2026-04-23-configure-and-make/
```

You might not have programmed in languages that needs compiling before, and instead used scripting languages like R or Python. Scripting languages requires that the one running the program has the language interpreter installed, i.e. to run a python script, you must have python installed.

```{.py filename="manual-compile/hello.py"}

> #!/bin/env python

Reviewer: ```suggestion
#!/usr/bin/env python3
```

print("Hello, World!")
```

and to run it:

```bash
python hello.py
```

> First we will cover a couple of basic steps when it comes to programming with compiled languages like C/C++. These basic steps is likely not something you will do when installing software that is created by someone else, but they will show why we will need to do the later steps.

Reviewer: ```suggestion
First we will cover a couple of basic steps when it comes to programming with compiled languages like C/C++. These basic steps are likely not something you will do when installing software that is created by someone else, but they will show why we will need to do the later steps.
```

### How to compile from source code

 We'll start from scratch with a simple C++ hello world program:

```{.cpp filename="manual-compile/hello.cpp"}
#include <iostream>

int main() {
    std::cout << "Hello, World!" << std::endl;
    return 0;
}
```

To be able to run this program we first have to compile it. Compiling is when you translate the human readable C++ code to binary machine code that the computer can run.

```{mermaid}
---
config:
  look: handDrawn
  theme: neutral
---
flowchart LR
    src("📄 hello.cpp<br/>(your code)")

>     cmd{{"g++ hellp.cpp -o hello"}}

Reviewer: ```suggestion
    cmd{{"g++ hello.cpp -o hello"}}
```

    bin("⚙️  hello<br/>(runnable!)")

    src --> cmd --> bin
```

```bash
# compile the code using the GNU C++ compiler and create the
# executable binary named `hello`
g++ hello.cpp -o hello

# run it from the terminal
./hello
Hello, World!
```

> If it is only a single `cpp` file that needs compiling, this manual approach is fine. However, more complex programs have multiple files that need compiling and some has to be compiled in the right order.

Reviewer: ```suggestion
If it is only a single `cpp` file that needs compiling, this manual approach is fine. However, more complex programs have multiple files that need compiling and some have to be compiled in the right order.
```

### Compiling using a Makefile

When a project grows to many files, typing every command by hand becomes tedious and error-prone. With a `Makefile` containing all the commands, just type `make` and it handles the rest.

```{mermaid}
---
config:

> ---

Reviewer: ```suggestion
```
No need for an empty config

flowchart LR
    cpps@{ shape: procs, label: "📄 .cpp files"}
    makefile("📒 Makefile")
    cmd{{"make"}}
    bin("⚙️  hello<br/>(runnable!)")

    cpps --> makefile --> cmd --> bin
```

Here is an example of how a simple `Makefile` can look:

```{.makefile filename="makefile-compile/Makefile"}
my_program: main.cpp engine.cpp utils.cpp
	g++ main.cpp engine.cpp utils.cpp -o my_program

clean:
	rm -f my_program

> ```

Reviewer: Can you use Code annotation to explain what each part is please. In comments perhaps have a bit with target: prerquisite; code. And how a target is only made if prerequisites are there, and if not will search for rule that describes how to make the prerequisite.

```{mermaid}
---
config:
  look: handDrawn
  theme: neutral
---
flowchart LR
    bsg("👷 Build system generator")
    makefile("📒 Makefile")
    cmd{{"make"}}
    bin("⚙️  hello<br/>(runnable!)")

    bsg --> makefile
    makefile --> cmd --> bin
```

### Prefix

Both Autotools and CMake can be tweaked by giving it options, and the option that controls where the software ends up when it is installed is called `prefix`. It is the path to the directory where all the compiled files will be copied to once the compilation is complete. This could be any folder you have write access to, such as your home directory or project directory.

For the remainder of the tutorial `$PREFIX` will point to `~/sw` to keep it short. Adjust it to suit your needs. If there are large programs you install you might fill up the quota in your home directory, so try to put it in project directories. Special consideration for NAISS systems, make sure that continuations of your project use the same *directory name* as the previous project, so that all paths remain the same.

```bash
# more realistic example

> #export PREFIX=/proj/naiss2026-99-999/software

Reviewer: ```suggestion
# export PREFIX=/proj/naiss2026-99-999/software
```

mkdir build
cd build
cmake .. -DCMAKE_INSTALL_PREFIX=$PREFIX -DCMAKE_BUILD_TYPE=Release
make -j4
make install
```

#### Example: Installing DIAMOND

DIAMOND is a sequence aligner for protein and translated DNA searches, designed for high performance analysis of big sequence data, and it is built using CMake.

```bash
cd ~/build

# download the source
wget https://github.com/bbuchfink/diamond/archive/refs/tags/v2.1.9.tar.gz
tar -xzf v2.1.9.tar.gz
cd diamond-2.1.9

# confirm it uses cmake
ls -l
# CMakeLists.txt  <-- yep

# create build dir
mkdir build && cd build

# create Makefile
cmake .. -DCMAKE_INSTALL_PREFIX=$PREFIX -DCMAKE_BUILD_TYPE=Release

# run Makefile
make -j4

# test that it works
./diamond --version

# install it
make install
```

## Making the tools available in your shell

Right now the binaries are in `$PREFIX/bin/`, but your shell doesn't know to look there. You need to add it to your `PATH`:

```bash
# add the bin directory to PATH for this session
export PREFIX=~/sw
export PATH=$PREFIX/bin:$PATH

# test that it works
which samtools
samtools --version

which diamond
diamond --version
```

To make this permanent, add it to your `~/.bashrc`: 

```bash

> # add these lines to ~/.bashrc and make sure

Reviewer: This comment is misleading. We want to run this in the terminal. Adding this to our `.bashrc` would lead to issues.

```suggestion
# Run these commands in your terminal to update your `~/.bashrc` file and make sure
```

---
title: "Introduction to Configure and Make"
description: "How to compile and install software in your own project directory on an HPC system, without needing root access or waiting for support staff."
author: "Martin Dahlö"
date: "2026-04-23"
categories:
  - HPC
  - Software installation
  - Bioinformatics
  - Terminal
  - Tips and Tricks
---

## Why compile your own software?

On a shared HPC system you typically don't have root access, which means you can't install software the usual way (e.g. `apt install` or `yum install`). Software that is needed for your research is usually installed centrally by support staff and made available via the module system (`module load`). This works well for widely used tools, but it can take time, and for more niche software it might never happen.

The good news is that you don't need root access to compile and install software! You just need to tell the installer to put the files somewhere you own, like your home directory or a project directory.

> An alternative to compiling the software is to check if the software is available through Docker, Conda/Pixi, or Apptainer. These approaches are more portable and often easier to get going with. This tutorial is aimed at exploring how to complile software if you have to. It is not because it is simpler than containerized solutions, it is often times not.

Reviewer: ```suggestion
::: {.callout-tip}

Before compiling from source, check if the software is available via Docker, Conda/Pixi, or Apptainer. These containerized/prepackaged approaches are more portable and easier to deploy. This tutorial focuses on the manual compilation process, which is often more complex but sometimes necessary.

:::
```

To be able to run this program we first have to compile it. Compiling is when you translate the human readable C++ code to binary machine code that the computer can run.

```{mermaid}
---
config:
  look: handDrawn
  theme: neutral
---
flowchart LR
    src("📄 hello.cpp<br/>(your code)")
    cmd{{"g++ hello.cpp -o hello"}}
    bin("⚙️  hello<br/>(runnable!)")

    src --> cmd --> bin
```

```bash
# compile the code using the GNU C++ compiler and create the
# executable binary named `hello`
g++ hello.cpp -o hello

# run it from the terminal
./hello
Hello, World!
```

If it is only a single `cpp` file that needs compiling, this manual approach is fine. However, more complex programs have multiple files that need compiling and some have to be compiled in the right order.

### Compiling using a Makefile

When a project grows to many files, typing every command by hand becomes tedious and error-prone. With a `Makefile` containing all the commands, just type `make` and it handles the rest.

```{mermaid}
flowchart LR
    cpps@{ shape: procs, label: "📄 .cpp files"}
    makefile("📒 Makefile")
    cmd{{"make"}}
    bin("⚙️  hello<br/>(runnable!)")

    cpps --> makefile --> cmd --> bin
```

> Here is an example of how a simple `Makefile` can look:
> 
> ```{.makefile filename="makefile-compile/Makefile"}
> # Rule to build the final executable "my_program"
> # The executable depends on these three source files
> my_program: main.cpp engine.cpp utils.cpp
> 	# Compile all source files and output the executable with -o flag
> 	g++ main.cpp engine.cpp utils.cpp -o my_program
> 
> # Rule to remove the compiled executable, helping to force a fresh build
> clean:
> 	# Remove "my_program" if it exists, the -f flag suppresses errors if it doesn't
> 	rm -f my_program
> ```

Reviewer: ````suggestion
```{.makefile filename="makefile-compile/Makefile"}
# target_file : source_files 
# <tab>commands

# Rule to build the final executable "my_program"
my_program: main.cpp engine.cpp utils.cpp # <1> <2>
	# Compile all source files and output the executable with -o flag
	g++ main.cpp engine.cpp utils.cpp -o my_program # <3> <4>

# Rule to remove the compiled executable, helping to force a fresh build
clean: # <5>
	# Remove "my_program" if it exists, the -f flag suppresses errors if it doesn't
	rm -f my_program
```

1. The executable `my_program` (target) depends on these three source files
2. `make my_program` (or `make` since it's the first rule) only runs if `my_program` doesn't exist, or the source files change.
3. The recipe is indented using a {{< kbd Tab >}} character ( leading spaces cause an error )
4. Each line runs in it's own subshell ( it doesn't know what's run before - like a directory change ).
5. The file `clean` doesn't exist, so the command `make clean` always runs the recipe.
````
